pytorch_lightning/loops/training_epoch_loop.py

```diff
@@ -0,0 +1,379 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Iterator, List, Union
+
+import pytorch_lightning as pl
+from pytorch_lightning.loops.base import Loop
+from pytorch_lightning.loops.training_batch_loop import TrainingBatchLoop
+from pytorch_lightning.trainer.connectors.logger_connector.result import ResultCollection
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.model_helpers import is_overridden
+from pytorch_lightning.utilities.signature_utils import is_param_in_hook_signature
+from pytorch_lightning.utilities.warnings import WarningCache
+
+
+class TrainingEpochLoop(Loop):
+    """ Runs over all batches in a dataloader (one epoch). """
+
+    def __init__(self, min_steps, max_steps):
+        super().__init__()
+        self.min_steps = min_steps
+        self.max_steps = max_steps
+
+        self.global_step = 0
+
+        # the total batch index across all epochs
+        self.total_batch_idx = 0
+        # the current batch index in the loop that runs over the dataloader(s)
+        self.iteration_count = 0
+        # the current split index when the batch gets split into chunks in truncated backprop through time
+        self.split_idx = None
+
+        self._dataloader_idx = None
+        self._should_stop = False
+
+        self.is_last_batch = None
+        self.batches_seen = 0
+        self.warning_cache = WarningCache()
+        self.epoch_output = None
+
+        self.batch_loop = None
+
+    @property
+    def batch_idx(self) -> int:
+        return self.iteration_count
+
+    @property
+    def done(self):
+        max_steps_reached = self.max_steps is not None and self.global_step >= self.max_steps
+        return max_steps_reached or self.trainer.should_stop or self._num_training_batches_reached(self.is_last_batch)
+
+    def connect(self, trainer: 'pl.Trainer', *args, **kwargs):
+        self.trainer = trainer
+        self.batch_loop = TrainingBatchLoop()
+        self.batch_loop.connect(trainer)
+
+    def run(self, *args, **kwargs):
+        self.reset()
+        self.on_run_start()
+
+        # TODO(@awaelchli): while condition is different from super.run(),
+        #   redesign the done conditions and use the base class run() implementation
+        while True:
+            try:
+                self.on_advance_start(*args, **kwargs)
+                self.advance(*args, **kwargs)
+                self.on_advance_end()
+                self.iteration_count += 1
+            except StopIteration:
+                break
```

Comment:
n00b question: large PR so its hard to see why the `done` condition here doesn't work? do we need to iterate at least once?

Reply:
removed!
we don't need this override anymore because we introduced the StopIteration in the base loop. 
There are two ways to exit the loop:
1) via done condition at while entry
2) via StopIteration at an arbitrary point